NUCLEO_L4R5ZI: add new platform

bcostm opened the pull request:
### Description

Add support of this board:
https://www.st.com/content/st_com/en/products/evaluation-tools/product-evaluation-tools/mcu-eval-tools/stm32-mcu-eval-tools/stm32-mcu-nucleo/nucleo-l4r5zi.html

All mbed-os-5 tests with ARM/GCC/IAR are PASS
[test_report.zip](https://github.com/ARMmbed/mbed-os/files/2222719/test_report.zip)



### Pull request type

    [ ] Fix
    [ ] Refactor
    [X] New target
    [ ] Feature
    [ ] Breaking change

bcostm committed:
NUCLEO_L4R5ZI: add all board files

bcostm committed:
NUCLEO_L4R5ZI: update mbed_rtx.h

bcostm committed:
NUCLEO_L4R5ZI: update targets.json

bcostm committed:
NUCLEO_L4R5ZI: update iar and sw4stm32 exporters

bcostm committed:
NUCLEO_L4R5ZI: add missing timer freeze macro

bcostm committed:
NUCLEO_L4R5ZI: enable bootloader

bcostm committed:
NUCLEO_L4R5ZI: remove PWM_5 pins as already used by us_ticker

bcostm committed:
NUCLEO_L4R5ZI: change _ALTx pins

bcostm committed:
NUCLEO_L4R5ZI: set IAR linker stack size to 1KB

0xc0170 commented:
/morph build

0xc0170 commented:
What version of IAR does this target require ? Our CI IAR version fails to find a target.

jeromecoutant commented:
Hi

> What version of IAR does this target require ? Our CI IAR version fails to find a target.

I don't really know IAR, but quick google search indicates version 8.20.1 ?
http://netstorage.iar.com/SuppDB/Public/UPDINFO/012495/arm/doc/infocenter/readme.ENU.html

0xc0170 commented:
if that is correct (that MCU definition is only for IAR 8.x) , the version we do not support it at the moment. That would explain the exporters failure

jeromecoutant commented:
@screamerbg

jeromecoutant commented:
@marcuschangarm

cmonr commented:
@ashok-rao

ashok-rao commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L4R5xI/TARGET_NUCLEO_L4R5ZI/PinNames.h:
```diff
@@ -0,0 +1,331 @@
+typedef enum {
+
+    PA_0  = 0x00,
+    PA_0_ALT0 = PA_0 | ALT0,
```
I like this a lot!! Opens up opportunity for multiplexing..

ashok-rao commented on targets/TARGET_STM/TARGET_STM32L4/TARGET_STM32L4R5xI/TARGET_NUCLEO_L4R5ZI/PinNames.h:
```diff
@@ -0,0 +1,331 @@
+    LED2        = PB_7, // Blue
+    LED3        = PB_14, // Red
+    LED4        = LED1,
+    LED_RED     = LED3,
```
Maybe create aliases for the Blue & Green as well (just for consistency..)?

0xc0170 commented:
If there is no IAR 7 support for this target, we might need to park this PR (close it for now).
It would be good to identify upcoming targets and their IAR 7 vs IAR 8 support.

bcostm commented:
I have launched all my tests using IAR V7.5 without any problem but the issue concerns the exporter only.

I searched also in IAR release notes and STM32L4R5/L4R7/L4R9/L4S5/L4S7/L4S9 devices have been added in IAR V8.20.1. Other devices like STM32L4R5xxx-P/L4R7xxx-P/L4R9xxx-P have been added in IAR V8.30.1.

Is the migration to IAR V8 planned on mbed soon ?

cmonr commented:
@bcostm We're definitely getting more pressure and requests on when we'll migrate, but as of right now, I'm unaware of anything planned.

For this PR, I think you might be able to do something similar for another target, in that device-specific exporting is disabled for IAR: https://github.com/ARMmbed/mbed-os/pull/7105#issuecomment-399198041

bcostm commented:
OK thanks for the hint. I'll do the same.

bcostm committed:
NUCLEO_L4R5ZI: remove IAR exporter

The STM32L4R5ZI device has been added in IAR V8.20.1

Remove the IAR exporter until IAR V8 is supported on mbed.

bcostm commented:
Rebased and removed the IAR exporter for now.

jeromecoutant commented:
"needs: work" label can be removed ?

cmonr commented:
/morph build

adbridge commented:
Looks like this is failing with uvision exporter:
```
startup_stm32l4r5xx.S: Error:  #5: cannot open source input file "__UVISION_VERSION SETA 521": No such file or directory
startup_stm32l4r5xx.S: __UVISION_VERSION SETA 521: 0 warnings, 1 errorstartup_stm32l4r5xx.S: Error: A1067E: Output file specified as 'mbed-os\targets\TARGET_STM\TARGET_STM32L4\TARGET_STM32L4R5xI\device\TOOLCHAIN_ARM_STD\startup_stm32l4r5xx.S', but it has already been specified as '.\build\startup_stm32l4r5xx.o'
startup_stm32l4r5xx.S: error: A1905U: Pre-processor step failed for '__UVISION_VERSION SETA 521'
```

cmonr commented:
I wonder...

@bcostm Does this need a new CMSIS pack installed for uVision? If so, could you provide a URL for it?

@adbridge Oddly enough, that error output generally indicates that a CMSIS pack is missing. It's weird.

bcostm commented:
> Does this need a new CMSIS pack installed for uVision? If so, could you provide a URL for it?

Surely as this device is quite "new". Did you check here ? https://www.keil.com/dd2/pack/ scroll down to the STMicroelectronics STM32L4 Series Device Support. The latest version is the 2.0.0

screamerbg commented:
@cmonr What is gating this PR? CMSIS pack is available on the link @bcostm provided. E.g.
```
 Version: 2.0.0 (2017-10-06) Keil.STM32L4xx_DFP.2.0.0.pack
 Download

This DFP requires STM32CubeMX V4.22.1 and STM32Cube_FW_L4 V1.9.0 or higher to be installed.

Reduced overall pack size by extracting only 'Drivers', 'Documentation' and 'Fonts' from Firmware Library pack.

Updated documentation for STM32Cube Framework usage.

    New projects require the component 'Device:STM32Cube Framework:STM32CubeMX' for configuration of pins and clocks.

Device support:

    Added STM32L4Rx, STM32L4Sx device support.
```

0xc0170 commented:
> @cmonr What is gating this PR? CMSIS pack is available on the link @bcostm provided. E.g.

The CI needs to be updated, will request an update (if it has not yet been done).

cc @OPpuolitaival

cmonr commented:
To make tracking a bit easier: IOTTESTINF-3857

0xc0170 commented:
The Ci update will be done the next week, we will restart this job once done. Sorry for the delay

screamerbg commented:
@0xc0170 This is unacceptable! How an issue with exporter CI is gating the acceptance of new MCU in Mbed OS? All comments were addressed on the 27th of August, which is exactly 1 mohth ago.

bcostm commented:
Rebased

cmonr commented:
@screamerbg Whenever uVision needs a new CMSIS pack, it needs to be installed manually in the Windows exporter VM. Because this modifies a CI VM, changes are managed  by the @ARMmbed/mbed-os-test team to make sure that the CI is kept stable.

As for the timing, this and other PRs got affected by 5.10 release activity. During the 5.10 release, we explicitly made sure to not update CI, which means we weren't able to apply the new CMSIS pack.

We're doing what we can to work through the backlog, like prioritizing the oldest PRs as well as PRs that carry a high importance. Feel free to suggest anything here or in emails.

bulislaw commented:
@bcostm @jeromecoutant sorry for the delay with this PR. We were in the release cycle, where the PRs targeting the release get priority and also we can't modify the CI during this time. I'll try to make sure that this is communicated better going forward, so you are aware why the PR is not progressing as smoothly as it should.

@ARMmbed/mbed-os-maintainers please prioritise this PR and merge it ASAP.

cmonr commented:
/morph build

cmonr commented:
Sorry, jumped the gun there. CI updates still need to be made.

0xc0170 commented:
I am checking today with @ARMmbed/mbed-os-test-team to get specific date/time for the update.

0xc0170 commented:
VM was updated,

/morph build

0xc0170 commented:
IAR nodes went offline, had to aborted it. I will restart soon, checking  if the nodes are back online

0xc0170 commented:
This new target passed uVision, we just need to fix CI exporters and restart the job 🙄

NirSonnenschein commented:
/morph export-build

NirSonnenschein commented:
/morph export-build

0xc0170 commented:
/morph export-build

0xc0170 commented:
/morph export-build

0xc0170 commented:
License issue, will restart shortly

0xc0170 commented:
/morph export-build
